---
output: 
    rmarkdown::md_document:
        variant: gfm
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.path = "inst/"
)
```

# sitePath: Phylogenetic pathwayâ€“dependent recognition of fixed substitutions and parallel mutations

## Getting help

Post on Bioconductor [support site](https://support.bioconductor.org/) if having trouble using `sitePath`. Or open an [issue](https://github.com/wuaipinglab/sitePath/issues/new?assignees=&labels=&template=bug_report.md&title=) if a bug is found.

## Installation

[R programming language](https://cran.r-project.org/) >= 4.1.0 is required to use `sitePath`.

The stable release is available on [Bioconductor](https://bioconductor.org/packages/sitePath/).
```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("sitePath")
```

The installation from [GitHub](https://github.com/wuaipinglab/sitePath/) is in experimental stage but gives the newest feature:
```r
if (!requireNamespace("remotes", quietly = TRUE))
    install.packages("remotes")

remotes::install_github("wuaipinglab/sitePath")
```


## Data preparation
You need a _tree_ and a _MSA_ (multiple sequence alignment) file and the sequence names have to be matched!
```{r data_prep}
library(sitePath) # Load the sitePath package

# The path to your tree and MSA files
tree_file <- system.file("extdata", "ZIKV.newick", package = "sitePath")
alignment_file <- system.file("extdata", "ZIKV.fasta", package = "sitePath")


tree <- read.tree(tree_file) # Read the tree file into R
align <- read.alignment(alignment_file, format = "fasta") # Read the MSA file into R

```

## Run analysis
Use `paraFixSites` function to find fixation and parallel sites.

```{r run_analysis}
options(list("cl.cores" = 1)) # Set this bigger than 1 to use multiprocessing

paraFix <- paraFixSites(tree, alignment = align) # Run analysis to find fixation and parallel sites
paraFix
```

## Fixation sites
Use `allSitesName` and set `type` as "fixation" to retrieve fixation sites name
``` {r fixSites_name}
allSitesName(paraFix, type = "fixation")
```

Use `plotFixationSites` to view fixation sites
```{r plot_fixSites}
plotFixationSites(paraFix) # View all fixation sites on the tree
plotFixationSites(paraFix, site = 139) # View a single site

```

## Parallel sites
Use `allSitesName` and set `type` as "parallel" to retrieve parallel sites name
``` {r paraSites_name}
allSitesName(paraFix, type = "parallel")
```

Use `plotParallelSites` to view parallel sites
```{r}
plotParallelSites(paraFix) # View all parallel sites on the tree
plotParallelSites(paraFix, site = 1118) # View a single site
```

## A more detailed demo

Click [here](https://wuaipinglab.github.io/sitePath/articles/sitePath.html) for a more detailed tutorial
