---
title: "Use sitePath to find fixation and parallel sites"
author: "Chengyang Ji"
package: sitePath
output:
  BiocStyle::html_document:
        toc_float: false
abstract: |
    In viral evolution, fixed substitutions in the nucleic acid or protein level are closely associated with maintaining viral function, while parallel mutation reflects the competitive nature in adaptive selection. The continued accumulation of large-scale viral sequence data enhances the challenge of identifying important mutations in a quick and accurate manner. In sitePath, the phylogenetic tree was separated into a set of inheritable phylogenetic pathways via an automated pathway-division method. Then, for each phylogenetic pathway, the identification of fixed substitutions was transformed into a local-optimal-solution problem directed by a minimal entropy algorithm. Finally, the parallel mutation was determined based on the recurrence of mutations among at least two phylogenetic pathways.
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Data preparation
You need a _tree_ and a _MSA_ (multiple sequence alignment) file and the sequence names have to be matched!
```{r data_prep}
library(sitePath) # Load the sitePath package

# The path to your tree and MSA files
tree_file <- system.file("extdata", "ZIKV.newick", package = "sitePath")
alignment_file <- system.file("extdata", "ZIKV.fasta", package = "sitePath")


tree <- read.tree(tree_file) # Read the tree file into R
align <- read.alignment(alignment_file, format = "fasta") # Read the MSA file into R

```

# Run analysis
Use `paraFixSites` function to find fixation and parallel sites.

```{r run_analysis}
options(list("cl.cores" = 1)) # Set this bigger than 1 to use multiprocessing

paraFix <- paraFixSites(tree, alignment = align) # Run analysis to find fixation and parallel sites
paraFix
```

# Visualization
Use `plotFixationSites` to view fixation sites
```{r plot_fixSites}
plotFixationSites(paraFix) # View all fixation sites on the tree
plotFixationSites(paraFix, site = 139) # View a single site

```
Use `plotParallelSites` to view parallel sites
```{r}
plotParallelSites(paraFix) # View all parallel sites on the tree
plotParallelSites(paraFix, site = 1118) # View a single site
```


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
